dali/pipeline/executor/workspace_policy.h

```diff
@@ -250,15 +252,188 @@ struct JIT_WS_Policy {
+          PlaceWorkspace<OpType::GPU, OpType::MIXED>(gpu_workspaces_, queue_idxs, graph,
+                                                     tensor_to_store_queue, mixed_op_stream,
+                                                     gpu_op_stream, mixed_op_events);
+        }
+      }
+    }
+  }
+
+  template <OpType op_type>
+  ws_t<op_type> GetWorkspace(QueueIdxs idxs, const OpGraph &graph, OpPartitionId partition_idx) {
+    // Handle MIXED and GPU indexing together
+    auto index_for_op_type = op_type == OpType::GPU ? OpType::MIXED : op_type;
+    int sequential_ws_idx = SequentialIndex(idxs, depths_, index_for_op_type);
+    auto &workspaces = GetWorkspacesCollection<op_type>();
+    return workspaces[sequential_ws_idx][partition_idx];
+  }
+
+  template <OpType op_type>
+  ws_t<op_type> GetWorkspace(QueueIdxs idxs, const OpGraph &graph, const OpNode &node) {
+    DALI_ENFORCE(node.op_type == op_type,
+                 "Wrong variant of method selected. OpType does not match.");
+    return GetWorkspace<op_type>(idxs, graph, node.partition_index);
+  }
+
+ private:
+  StageQueues depths_;
+  // ws_id -> op_id -> workspace
+  std::vector<std::vector<SupportWorkspace>> support_workspaces_;
+  std::vector<std::vector<HostWorkspace>> cpu_workspaces_;
+  std::vector<std::vector<MixedWorkspace>> mixed_workspaces_;
+  std::vector<std::vector<DeviceWorkspace>> gpu_workspaces_;
+
+  template <OpType op_type>
+  using ws_collection_t = typename std::vector<std::vector<op_type_to_workspace_t<op_type>>>;
+
+  // Access one of `support_workspaces_`, `cpu_workspaces_`, `mixed_workspaces_`, `gpu_workspaces_`
+  // based on op_type. Below are specialization for the sake of simplicity
+  template <OpType op_type>
+  ws_collection_t<op_type> &GetWorkspacesCollection() {
+    return detail::get<ws_collection_t<op_type>&>(
+        std::tie(support_workspaces_, cpu_workspaces_, mixed_workspaces_, gpu_workspaces_));
+  }
+
+  template <OpType op_type, OpType group_as = op_type, typename T>
+  void PlaceWorkspace(T &workspaces, QueueIdxs idxs, const OpGraph &graph,
+                      const std::vector<tensor_data_store_queue_t> &tensor_to_store_queue,
+                      cudaStream_t mixed_op_stream, cudaStream_t gpu_op_stream,
+                      const std::vector<std::vector<cudaEvent_t>> &mixed_op_events) {
+    int sequential_ws_idx = SequentialIndex(idxs, depths_, group_as);
+    workspaces[sequential_ws_idx].resize(graph.NumOp(op_type));
+    for (OpPartitionId partition_idx = 0; partition_idx < graph.NumOp(op_type); partition_idx++) {
+      auto &node = graph.Node(op_type, partition_idx);
+      workspaces[sequential_ws_idx][partition_idx] =
+          CreateWorkspace<op_type>(graph, node, tensor_to_store_queue, mixed_op_stream,
+                                   gpu_op_stream, mixed_op_events, idxs);
+    }
+  }
+
+  static_assert(std::is_same<QueuePolicy, SeparateQueuePolicy>::value,
```

Comment:
This is silly. Why is this class even a template, if it only accepts _one_ argument? Either make it a specialization of `AOT_WS_Policy` or make both `AOT_WS_Policy` and `AOT_Sep_WS_Policy` non-templates.

Reply:
Yeah, the specialization is good idea.

Reply:
Done